coderd/rbac/USAGE.md

```diff
@@ -0,0 +1,344 @@
+        // Auditors should be able to read any other user's frobulators
+        // Owner should be able to read any other user's frobulators
+        Name:     "FrobulatorsReadOnly",
+        Actions:  []policy.Action{policy.ActionRead},
+        Resource: rbac.ResourceFrobulator.WithOwner(currentUser.String()).InOrg(orgID),
+        AuthorizeMap: map[bool][]hasAuthSubjects{
+            true:  {orgMemberMe, orgAdmin, owner, orgAuditor},
+            false: {setOtherOrg, memberMe, templateAdmin, userAdmin, orgTemplateAdmin, orgUserAdmin},
+        },
+    },
+```
+
+Note how the `FrobulatorsModify` test case is just validating the
+`policy.ActionCreate, policy.ActionUpdate, policy.ActionDelete` actions, and
+only the **orgMember**, **orgAdmin**, and **owner** can access it.
+
+Similarly, the `FrobulatorsReadOnly` test case is only validating
+`policy.ActionRead`, which is allowed on all of the above plus the
+**orgAuditor** role.
+
+Now the tests pass, because we have covered all the possible scenarios:
+
+```bash
+$ go test github.com/coder/coder/v2/coderd/rbac -count=1
+ok  	github.com/coder/coder/v2/coderd/rbac	1.313s
+```
+
+When a case is not covered, you'll see an error like this (I moved the
+`orgAuditor` option from `true` to `false):
+
+```bash
+--- FAIL: TestRolePermissions (0.79s)
+    --- FAIL: TestRolePermissions/FrobulatorsReadOnly (0.01s)
+        roles_test.go:737:
+            	Error Trace:	/tmp/coder/coderd/rbac/roles_test.go:737
+            	Error:      	An error is expected but got nil.
+            	Test:       	TestRolePermissions/FrobulatorsReadOnly
+            	Messages:   	Should fail: FrobulatorsReadOnly as "org_auditor" doing "read" on "frobulator"
+FAIL
+FAIL	github.com/coder/coder/v2/coderd/rbac	1.390s
+FAIL
+```
+
+This shows you that the `org_auditor` role has `read` permissions on the
+frobulator, but no test case covered it.
+
+**NOTE: don't just add cases which make the tests pass; consider all the way in
+which your resource must be used, and test all of those scenarios!**
+
+# Database authorization
+
+Now that we have the RBAC system fully configured, we need to make use of it.
+
+Let's add a SQL query to `coderd/database/queries/frobulators.sql`:
+
+```sql
+-- name: GetUserFrobulators :many
+SELECT *
+FROM frobulators
+WHERE user_id = @user_id::uuid;
+```
+
+Once we run `make gen`, we'll find some stubbed code in
+`coderd/database/dbauthz/dbauthz.go`.
+
+```go
+...
+func (q *querier) GetUserFrobulators(ctx context.Context) ([]database.Frobulator, error) {
+    panic("not implemented")
+}
+...
+```
+
+Let's modify this function:
+
+```go
+...
+func (q *querier) GetUserFrobulators(ctx context.Context, userID uuid.UUID) ([]database.Frobulator, error) {
+	if err := q.authorizeContext(ctx, policy.ActionRead, rbac.ResourceFrobulator.WithOwner(userID.String())); err != nil {
+		return nil, err
+	}
+	return q.db.GetUserFrobulators(ctx, userID)
+}
+...
+```
```

Comment:
This is correct, except `rbac.ResourceFrobulator.WithOwner(userID.String())` should also include an `InOrg(...)` and even an `WithID(...)` to be 100% correct.

Fetching a resource is a super common pattern, and unfortunately you generally need to have the object in hand to know if you can read it. So, the generic function `fetch(...)` handles all of this **iff** you implement `RBACObject()` (which you should!).

```suggestion
...
func (q *querier) GetUserFrobulators(ctx context.Context, userID uuid.UUID) ([]database.Frobulator, error) {
  return fetch(q.log, q.auth, q.db.GetUserFrobulators)(ctx, id)
}
...
```
```